File: frontend/src/api/handbook.ts
// src/api/handbook.ts
import { client } from "./client";

// ── Types ──────────────────────────────────────────────────────────────────────

export interface ResourceLink {
  url: string;
  title: string;
  resource_type: "link" | "document" | "regulation" | "video";
}

export interface ChangeLogEntry {
  id: string;
  entity_type: string;
  changed_by: string;
  action: "created" | "updated" | "deleted";
  field_name?: string;
  old_value?: string;
  new_value?: string;
  diff?: Record<string, { old: unknown; new: unknown }>;
  created_at: string;
}

export interface NewsLink {
  id: string;
  excerpt?: string;
  article_id?: string;
  generated_news_id?: string;
  entity_type: string;
  country_id?: string;
  org_unit_id?: string;
  person_id?: string;
  note?: string;
  pinned_by?: string;
  created_at: string;
}
export interface HandbookEvent {
  id: string;
  person_id?: string | null;
  org_unit_id?: string | null;
  country_id?: string | null;
  title: string;
  event_type: string;
  date: string; // ISO datetime
  location?: string | null;
  description?: string | null;
  participants: string[];
  source_url?: string | null;
  article_id?: string | null;
  generated_news_id?: string | null;
  created_by?: string | null;
  created_at: string;
  updated_at: string;
}

export interface EventCreatePayload {
  person_id?: string;
  org_unit_id?: string;
  country_id?: string;
  title: string;
  event_type: string;
  date: string; // ISO datetime
  location?: string;
  description?: string;
  participants?: string[];
  source_url?: string;
  article_id?: string;
  generated_news_id?: string;
}

export interface Person {
  id: string;
  org_unit_id?: string;
  country_id: string;
  first_name: string;

  last_name: string;
  patronymic?: string;
  position_title?: string;
  rank?: string;
  photo_url?: string;
  bio?: string;
  contacts: Record<string, string>;
  resources: ResourceLink[];
  date_appointed?: string;
  date_dismissed?: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  news_links: NewsLink[];
  changelog: ChangeLogEntry[];
}

export interface OrgUnit {
  id: string;
  country_id: string;
  parent_id?: string;
  leader_title?: string;
  leader: Person | null;
  name: string;
  short_name?: string;
  unit_type: string;
  level: number;
  sort_order: number;
  description?: string;
  legal_basis?: string;
  resources: ResourceLink[];
  is_active: boolean;
  valid_from?: string;
  valid_to?: string;
  created_at: string;
  updated_at: string;
  children: OrgUnit[];
  persons: Person[];
  news_links: NewsLink[];
  changelog: ChangeLogEntry[];
}

export interface Country {
  id: string;
  code: string;
  name_uk: string;
  name_en: string;
  flag_emoji?: string;
  capital?: string;
  description?: string;
  resources: ResourceLink[];
  is_active: boolean;
  created_at: string;
  updated_at: string;
  org_units_count: number;
  persons_count: number;
}

export interface CountryDetail extends Country {
  org_units: OrgUnit[];
  changelog: ChangeLogEntry[];
  news_links: NewsLink[];
}

export interface SearchResult {
  entity_type: "country" | "org_unit" | "person";
  id: string;
  title: string;
  subtitle?: string;
  country_code?: string;
  country_name?: string;
}

export interface PaginatedCountries {
  items: Country[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

// ── Helper ─────────────────────────────────────────────────────────────────────
export function buildTree(flatUnits: OrgUnit[]): OrgUnit[] {
  const map = new Map<string, OrgUnit>();
  const roots: OrgUnit[] = [];

  // Clone to avoid mutating
  const units = flatUnits.map((u) => ({ ...u, children: [] }));
  units.forEach((u) => map.set(u.id, u));
  units.forEach((u) => {
    if (u.parent_id && map.has(u.parent_id)) {
      map.get(u.parent_id)!.children.push(u);
    } else {
      roots.push(u);
    }
  });
  return roots;
}

export function fullName(
  p: Pick<Person, "last_name" | "first_name" | "patronymic">,
): string {
  return [p.last_name, p.first_name, p.patronymic].filter(Boolean).join(" ");
}

// ── API calls ──────────────────────────────────────────────────────────────────

const BASE = "/handbook";
export interface Event {
  id: string;
  person_id?: string;
  org_unit_id?: string;
  country_id?: string;
  title: string;
  event_type: string;
  date: string;
  location?: string;
  description?: string;
  participants: string[];
  source_url?: string;
  article_id?: string;
  generated_news_id?: string;
  created_by?: string;
  created_at: string;
  updated_at: string;
}
export interface EventCreate {
  person_id?: string;
  org_unit_id?: string;
  country_id?: string;
  title: string;
  event_type: string;
  date: string;
  location?: string;
  description?: string;
  participants?: string[];
  source_url?: string;
  article_id?: string;
  generated_news_id?: string;
}
// Оновлений NewsLink з excerpt
export interface NewsLinkWithExcerpt extends NewsLink {
  excerpt?: string; // виділений фрагмент тексту статті
}
export const handbookApi = {
  // Countries
  listCountries: async (params?: {
    q?: string;
    page?: number;
    page_size?: number;
  }) => {
    const { data } = await client.get<PaginatedCountries>(`${BASE}/countries`, {
      params,
    });
    return data;
  },
  getPerson: async (id: string) => {
    const { data } = await client.get<Person>(`${BASE}/persons/${id}`);
    return data;
  },
  getPersonEvents: async (personId: string): Promise<HandbookEvent[]> => {
    const res = await fetch(`/api/handbook/persons/${personId}/events`);
    if (!res.ok) throw new Error("Failed to load person events");
    return res.json();
  },

  /** Список заходів підрозділу */
  getOrgUnitEvents: async (orgUnitId: string): Promise<HandbookEvent[]> => {
    const res = await fetch(`/api/handbook/org-units/${orgUnitId}/events`);
    if (!res.ok) throw new Error("Failed to load org unit events");
    return res.json();
  },

  /** Список заходів країни */
  getCountryEvents: async (countryId: string): Promise<HandbookEvent[]> => {
    const res = await fetch(`/api/handbook/countries/${countryId}/events`);
    if (!res.ok) throw new Error("Failed to load country events");
    return res.json();
  },

  /** Створити захід */
  createEvent: async (
    payload: EventCreatePayload,
    editor = "user",
  ): Promise<HandbookEvent> => {
    const res = await fetch("/api/handbook/events", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-Changed-By": editor,
      },
      body: JSON.stringify(payload),
    });
    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err?.detail ?? "Failed to create event");
    }
    return res.json();
  },

  /** Оновити захід */
  updateEvent: async (
    eventId: string,
    payload: Partial<EventCreatePayload>,
  ): Promise<HandbookEvent> => {
    const res = await fetch(`/api/handbook/events/${eventId}`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err?.detail ?? "Failed to update event");
    }
    return res.json();
  },

  /** Видалити захід */
  deleteEvent: async (eventId: string): Promise<void> => {
    const res = await fetch(`/api/handbook/events/${eventId}`, {
      method: "DELETE",
    });
    if (!res.ok) throw new Error("Failed to delete event");
  },

  getArticleEvents: async (articleId: string) => {
    const { data } = await client.get<Event[]>(
      `${BASE}/articles/${articleId}/events`,
    );
    return data;
  },

  getCountry: async (id: string) => {
    const { data } = await client.get<CountryDetail>(`${BASE}/countries/${id}`);
    return data;
  },
  createCountry: async (body: Partial<Country>, changedBy = "user") => {
    const { data } = await client.post<Country>(`${BASE}/countries`, body, {
      headers: { "X-Changed-By": changedBy },
    });
    return data;
  },
  getLinksForGeneratedNews: async (generatedNewsId: string) => {
    const { data } = await client.get<NewsLink[]>(
      `${BASE}/news-links/generated-news/${generatedNewsId}`,
    );
    return data;
  },
  updateCountry: async (
    id: string,
    body: Partial<Country>,
    changedBy = "user",
  ) => {
    const { data } = await client.patch<Country>(
      `${BASE}/countries/${id}`,
      body,
      {
        headers: { "X-Changed-By": changedBy },
      },
    );
    return data;
  },
  deleteCountry: async (id: string, changedBy = "user") => {
    await client.delete(`${BASE}/countries/${id}`, {
      headers: { "X-Changed-By": changedBy },
    });
  },

  // Org Units
  getOrgTree: async (countryId: string) => {
    const { data } = await client.get<OrgUnit[]>(
      `${BASE}/countries/${countryId}/org-units`,
    );
    return data;
  },
  createOrgUnit: async (body: Partial<OrgUnit>, changedBy = "user") => {
    const { data } = await client.post<OrgUnit>(`${BASE}/org-units`, body, {
      headers: { "X-Changed-By": changedBy },
    });
    return data;
  },
  updateOrgUnit: async (
    id: string,
    body: Partial<OrgUnit>,
    changedBy = "user",
  ) => {
    const { data } = await client.patch<OrgUnit>(
      `${BASE}/org-units/${id}`,
      body,
      {
        headers: { "X-Changed-By": changedBy },
      },
    );
    return data;
  },
  moveOrgUnit: async (
    id: string,
    newParentId: string | null,
    changedBy = "user",
  ) => {
    const { data } = await client.post<OrgUnit>(
      `${BASE}/org-units/${id}/move`,
      null,
      {
        params: { new_parent_id: newParentId },
        headers: { "X-Changed-By": changedBy },
      },
    );
    return data;
  },
  deleteOrgUnit: async (id: string, changedBy = "user") => {
    await client.delete(`${BASE}/org-units/${id}`, {
      headers: { "X-Changed-By": changedBy },
    });
  },

  // Persons
  listPersons: async (countryId: string) => {
    const { data } = await client.get<Person[]>(
      `${BASE}/countries/${countryId}/persons`,
    );
    return data;
  },
  createPerson: async (body: Partial<Person>, changedBy = "user") => {
    const { data } = await client.post<Person>(`${BASE}/persons`, body, {
      headers: { "X-Changed-By": changedBy },
    });
    return data;
  },
  updatePerson: async (
    id: string,
    body: Partial<Person>,
    changedBy = "user",
  ) => {
    const { data } = await client.patch<Person>(`${BASE}/persons/${id}`, body, {
      headers: { "X-Changed-By": changedBy },
    });
    return data;
  },
  deletePerson: async (id: string, changedBy = "user") => {
    await client.delete(`${BASE}/persons/${id}`, {
      headers: { "X-Changed-By": changedBy },
    });
  },

  // News Links
  createNewsLink: async (
    body: Partial<NewsLinkWithExcerpt> & { excerpt?: string },
    changedBy = "user",
  ) => {
    const { data } = await client.post<NewsLinkWithExcerpt>(
      `${BASE}/news-links`,
      body,
      {
        headers: { "X-Changed-By": changedBy },
      },
    );
    return data;
  },
  getLinksForArticle: async (articleId: string) => {
    const { data } = await client.get<NewsLink[]>(
      `${BASE}/news-links/article/${articleId}`,
    );
    return data;
  },
  getLinksForEntity: async (entityType: string, entityId: string) => {
    const { data } = await client.get<NewsLink[]>(
      `${BASE}/news-links/${entityType}/${entityId}`,
    );
    return data;
  },
  deleteNewsLink: async (linkId: string) => {
    await client.delete(`${BASE}/news-links/${linkId}`);
  },

  // Search
  search: async (q: string) => {
    const { data } = await client.get<{
      query: string;
      total: number;
      items: SearchResult[];
    }>(`${BASE}/search`, { params: { q } });
    return data;
  },
};

File: backend/src/presentation/api/routes/handbook.py
# src/api/routers/handbook.py
"""
FastAPI router — Handbook API.

Mount in your main app:
    from src.api.routers.handbook import router as handbook_router
    app.include_router(handbook_router, prefix="/handbook", tags=["handbook"])

All write operations accept X-Changed-By header (username of editor).
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Header, Query
from sqlalchemy.ext.asyncio import AsyncSession

from src.config.container import get_container

async def get_db_session():
    """
    Залежність FastAPI, яка бере сесію з нашого Singleton-контейнера
    та гарантує правильний lifecycle транзакції (commit/rollback).
    """
    container = get_container()
    async with container.db_session() as session:
        yield session
        
from src.infrastructure.persistence.repositories.handbook_repo import (
    CountryRepository,
    OrgUnitRepository,
    PersonRepository,
    NewsLinkRepository,
    HandbookSearchRepository,
)
from src.presentation.api.schemas.handbook_schemas import (
    CountryCreate, CountryUpdate, CountryOut, CountryDetail, PaginatedCountries,
    OrgUnitCreate, OrgUnitUpdate, OrgUnitOut,
    PersonCreate, PersonUpdate, PersonOut,
    NewsLinkCreate, NewsLinkOut,
    SearchResponse, SearchResult,
)

router = APIRouter()


def _editor(x_changed_by: str = Header(default="anonymous")) -> str:
    return x_changed_by


# ══════════════════════════════════════════════════════════════════════════════
# Countries
# ══════════════════════════════════════════════════════════════════════════════

@router.get("/countries", response_model=PaginatedCountries)
async def list_countries(
    q: str | None = Query(None),
    is_active: bool | None = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(30, ge=1, le=100),
    session: AsyncSession = Depends(get_db_session),
):
    repo = CountryRepository(session)
    offset = (page - 1) * page_size
    items, total = await repo.list(q=q, is_active=is_active, limit=page_size, offset=offset)
    pages = (total + page_size - 1) // page_size
    return PaginatedCountries(
        items=items, total=total, page=page, page_size=page_size, pages=pages
    )


@router.post("/countries", response_model=CountryOut, status_code=201)
async def create_country(
    body: CountryCreate,
    editor: str = Depends(_editor),
    session: AsyncSession = Depends(get_db_session),
):
    repo = CountryRepository(session)
    existing = await repo.get_by_code(body.code)
    if existing:
        raise HTTPException(409, f"Country with code '{body.code}' already exists")
    data = body.model_dump()
    m = await repo.create(data, editor)
    await session.commit()
    return m


@router.get("/countries/{country_id}", response_model=CountryDetail)
async def get_country(
    country_id: str,
    session: AsyncSession = Depends(get_db_session),
):
    repo = CountryRepository(session)
    m = await repo.get_detail(country_id)
    if not m:
        raise HTTPException(404, "Country not found")
    return m


@router.patch("/countries/{country_id}", response_model=CountryOut)
async def update_country(
    country_id: str,
    body: CountryUpdate,
    editor: str = Depends(_editor),
    session: AsyncSession = Depends(get_db_session),
):
    repo = CountryRepository(session)
    data = body.model_dump(exclude_none=True)
    m = await repo.update(country_id, data, editor)
    if not m:
        raise HTTPException(404, "Country not found")
    await session.commit()
    return m


@router.delete("/countries/{country_id}", status_code=204)
async def delete_country(
    country_id: str,
    editor: str = Depends(_editor),
    session: AsyncSession = Depends(get_db_session),
):
    repo = CountryRepository(session)
    ok = await repo.delete(country_id, editor)
    if not ok:
        raise HTTPException(404, "Country not found")
    await session.commit()


# ══════════════════════════════════════════════════════════════════════════════
# Org Units
# ══════════════════════════════════════════════════════════════════════════════

@router.get("/countries/{country_id}/org-units", response_model=list[OrgUnitOut])
async def get_org_tree(
    country_id: str,
    session: AsyncSession = Depends(get_db_session),
):
    """Return flat list of all org units; tree structure built on frontend."""
    repo = OrgUnitRepository(session)
    items = await repo.get_tree(country_id)
    return items


@router.post("/org-units", response_model=OrgUnitOut, status_code=201)
async def create_org_unit(
    body: OrgUnitCreate,
    editor: str = Depends(_editor),
    session: AsyncSession = Depends(get_db_session),
):
    repo = OrgUnitRepository(session)
    data = body.model_dump()
    m = await repo.create(data, editor)
    
    # 1. Flush to push changes to the DB and get the ID (keeps transaction open)
    await session.commit() 
    
    # 2. Re-fetch to load relationships (children, persons, etc.)
    return await repo.get(m.id)

@router.get("/org-units/{unit_id}", response_model=OrgUnitOut)
async def get_org_unit(
    unit_id: str,
    session: AsyncSession = Depends(get_db_session),
):
    repo = OrgUnitRepository(session)
    m = await repo.get(unit_id)
    if not m:
        raise HTTPException(404, "OrgUnit not found")
    return m


@router.patch("/org-units/{unit_id}", response_model=OrgUnitOut)
async def update_org_unit(
    unit_id: str,
    body: OrgUnitUpdate,
    editor: str = Depends(_editor),
    session: AsyncSession = Depends(get_db_session),
):
    repo = OrgUnitRepository(session)
    data = body.model_dump(exclude_none=True)
    m = await repo.update(unit_id, data, editor)
    if not m:
        raise HTTPException(404, "OrgUnit not found")
        
    # Flush instead of commit
    await session.commit()  
    
    return await repo.get(unit_id)


@router.post("/org-units/{unit_id}/move", response_model=OrgUnitOut)
async def move_org_unit(
    unit_id: str,
    new_parent_id: str | None = Query(None),
    editor: str = Depends(_editor),
    session: AsyncSession = Depends(get_db_session),
):
    repo = OrgUnitRepository(session)
    m = await repo.move(unit_id, new_parent_id, editor)
    if not m:
        raise HTTPException(404, "OrgUnit not found")
        
    # Flush instead of commit
    await session.commit()
    
    return await repo.get(unit_id)


@router.delete("/org-units/{unit_id}", status_code=204)
async def delete_org_unit(
    unit_id: str,
    editor: str = Depends(_editor),
    session: AsyncSession = Depends(get_db_session),
):
    repo = OrgUnitRepository(session)
    ok = await repo.delete(unit_id, editor)
    if not ok:
        raise HTTPException(404, "OrgUnit not found")
    await session.commit()


# ══════════════════════════════════════════════════════════════════════════════
# Persons
# ══════════════════════════════════════════════════════════════════════════════

@router.get("/countries/{country_id}/persons", response_model=list[PersonOut])
async def list_persons(
    country_id: str,
    session: AsyncSession = Depends(get_db_session),
):
    repo = PersonRepository(session)
    return await repo.list_by_country(country_id)


@router.post("/persons", response_model=PersonOut, status_code=201)
async def create_person(
    body: PersonCreate,
    editor: str = Depends(_editor),
    session: AsyncSession = Depends(get_db_session),
):
    repo = PersonRepository(session)
    data = body.model_dump()
    m = await repo.create(data, editor)
    await session.commit()
    return m


@router.get("/persons/{person_id}", response_model=PersonOut)
async def get_person(
    person_id: str,
    session: AsyncSession = Depends(get_db_session),
):
    repo = PersonRepository(session)
    m = await repo.get(person_id)
    if not m:
        raise HTTPException(404, "Person not found")
    return m


@router.patch("/persons/{person_id}", response_model=PersonOut)
async def update_person(
    person_id: str,
    body: PersonUpdate,
    editor: str = Depends(_editor),
    session: AsyncSession = Depends(get_db_session),
):
    repo = PersonRepository(session)
    data = body.model_dump(exclude_none=True)
    m = await repo.update(person_id, data, editor)
    if not m:
        raise HTTPException(404, "Person not found")
    await session.commit()
    return m


@router.delete("/persons/{person_id}", status_code=204)
async def delete_person(
    person_id: str,
    editor: str = Depends(_editor),
    session: AsyncSession = Depends(get_db_session),
):
    repo = PersonRepository(session)
    ok = await repo.delete(person_id, editor)
    if not ok:
        raise HTTPException(404, "Person not found")
    await session.commit()

# ✅ Додати в handbook.py
@router.get("/news-links/generated-news/{generated_news_id}", response_model=list[NewsLinkOut])
async def get_links_for_generated_news(
    generated_news_id: str,
    session: AsyncSession = Depends(get_db_session),
):
    repo = NewsLinkRepository(session)
    return await repo.list_for_generated_news(generated_news_id)
# ══════════════════════════════════════════════════════════════════════════════
# News Links
# ══════════════════════════════════════════════════════════════════════════════

@router.post("/news-links", response_model=NewsLinkOut, status_code=201)
async def create_news_link(
    body: NewsLinkCreate,
    editor: str = Depends(_editor),
    session: AsyncSession = Depends(get_db_session),
):
    repo = NewsLinkRepository(session)
    data = body.model_dump()
    data["pinned_by"] = editor
    m = await repo.create(data)
    await session.commit()
    return m


@router.get("/news-links/article/{article_id}", response_model=list[NewsLinkOut])
async def get_links_for_article(
    article_id: str,
    session: AsyncSession = Depends(get_db_session),
):
    repo = NewsLinkRepository(session)
    return await repo.list_for_article(article_id)


@router.get("/news-links/{entity_type}/{entity_id}", response_model=list[NewsLinkOut])
async def get_links_for_entity(
    entity_type: str,
    entity_id: str,
    session: AsyncSession = Depends(get_db_session),
):
    repo = NewsLinkRepository(session)
    return await repo.list_for_entity(entity_type, entity_id)


@router.delete("/news-links/{link_id}", status_code=204)
async def delete_news_link(
    link_id: str,
    session: AsyncSession = Depends(get_db_session),
):
    repo = NewsLinkRepository(session)
    ok = await repo.delete(link_id)
    if not ok:
        raise HTTPException(404, "NewsLink not found")
    await session.commit()


# ══════════════════════════════════════════════════════════════════════════════
# Search
# ══════════════════════════════════════════════════════════════════════════════

@router.get("/search", response_model=SearchResponse)
async def search_handbook(
    q: str = Query(min_length=2),
    limit: int = Query(20, ge=1, le=50),
    session: AsyncSession = Depends(get_db_session),
):
    repo = HandbookSearchRepository(session)
    items = await repo.search(q, limit=limit)
    return SearchResponse(
        query=q,
        total=len(items),
        items=[SearchResult(**i) for i in items],
    )